tests/integration/managed_shared_ptr_tests.cpp

```diff
@@ -0,0 +1,360 @@
+  CHAI_HOST_DEVICE virtual void d_function(void) const = 0;
+  CHAI_HOST_DEVICE virtual void set_content(unsigned long long) = 0;
+};
+
+class A2
+{
+public:
+  CHAI_HOST_DEVICE A2(void) { printf("++ A2 has been constructed\n"); }
+  CHAI_HOST_DEVICE ~A2(void) { printf("-- A2 has been destructed\n"); }
+};
+
+class B final : public A, public A2
+{
+public:
+  unsigned long long content_B;
+  CHAI_HOST_DEVICE B(void) : content_B(0xBBBBBBBBBBBBBBBBull) { printf("++ B has been constructed\n"); }
+  CHAI_HOST_DEVICE ~B(void) { printf("-- B has been destructed\n"); }
+  CHAI_HOST_DEVICE virtual void function(void) const override { printf("%llX\n", content_B); }
+  CHAI_HOST_DEVICE virtual void d_function(void) const override { d.function(); }
+  CHAI_HOST_DEVICE virtual void set_content(unsigned long long val) override { content_B = val; content_A = val; }
+};
+
+
+class AAbsMem : public chai::CHAICopyable , public chai::expt::CHAIPoly
+{
+public:
+  chai::expt::ManagedSharedPtr<const C> base_member;
+  unsigned long long content_A;
+
+  CHAI_HOST_DEVICE AAbsMem(void) : content_A(0xAAAAAAAAAAAAAAAAull) { printf("++ A has been constructed\n"); }
+
+  //template<typename Derived, typename = typename std::enable_if<std::is_base_of<C, Derived>::value>::type >
+  template<typename Derived>
+  CHAI_HOST AAbsMem(Derived const& base_val)
+    //: base_member(chai::expt::make_shared<const Derived>(base_val))
+    : base_member(chai::expt::make_shared<Derived>(base_val))
+    , content_A(0xAAAAAAAAAAAAAAAAull) 
+  { printf("++ A has been constructed\n"); }
+
+  CHAI_HOST_DEVICE virtual ~AAbsMem(void) { printf("-- A has been destructed\n"); }
+  CHAI_HOST_DEVICE virtual void function(void) const = 0;
+  CHAI_HOST_DEVICE virtual void d_function(void) const = 0;
+  CHAI_HOST_DEVICE virtual void set_content(unsigned long long) = 0;
+};
+
+class BAbsMem final : public AAbsMem
+{
+public:
+  unsigned long long content_B;
+
+  CHAI_HOST_DEVICE BAbsMem() : AAbsMem()
+  { 
+    printf("++ B has been constructed\n"); 
+  }
+
+  template<typename Derived>
+  CHAI_HOST BAbsMem(Derived const& base_val) 
+    : AAbsMem(base_val)
+    , content_B(0xBBBBBBBBBBBBBBBBull) 
+  { 
+    printf("++ B has been constructed\n"); 
+  }
+
+  CHAI_HOST_DEVICE ~BAbsMem(void) { printf("-- B has been destructed\n"); }
+  CHAI_HOST_DEVICE virtual void function(void) const override { printf("%llX\n", content_B); }
+  CHAI_HOST_DEVICE virtual void d_function(void) const override { base_member->function(); }
+  CHAI_HOST_DEVICE virtual void set_content(unsigned long long val) override { content_B = val; content_A = val; }
+};
+
+class NV
+{
+public:
+  unsigned long long content_NV;
+  CHAI_HOST_DEVICE NV(void) : content_NV(0xFFFFFFFFFFFFFFFFull) { printf("++ NV has been constructed\n"); }
+  CHAI_HOST_DEVICE ~NV(void) { printf("-- NV has been destructed\n"); }
+  CHAI_HOST_DEVICE void function(void) const { printf("%llX\n", content_NV); }
+};
+
+GPU_TEST(managed_shared_ptr, shared_ptr_absmem)
```

Comment:
Since these tests are a little more complicated and the variable naming is not clear to me, could you provide some brief documentation for each test case?